File: platform/backend/app/api/content.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (c) 2026 CommonHuman-Lab
from datetime import datetime
from typing import Any, Optional

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db, require_admin
from app.models.content import ContentStatus, ContentType, ReviewVerdict
from app.models.user import User
from app.services.content_service import (
    claim_review, create_submission, get_submission_or_404,
    list_submissions, publish_submission, record_review,
    require_platform_role, submit_for_review, update_submission,
)

router = APIRouter(prefix="/content", tags=["content"])


# ── Schemas ───────────────────────────────────────────────────────────────────

class SubmissionOut(BaseModel):
    id: int
    author_id: int
    author_username: Optional[str] = None
    content_type: ContentType
    title: str
    body: dict[str, Any]
    status: ContentStatus
    reviewer_id: Optional[int]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


def _sub_out(sub) -> SubmissionOut:
    return SubmissionOut(
        id=sub.id,
        author_id=sub.author_id,
        author_username=sub.author.username if sub.author else None,
        content_type=sub.content_type,
        title=sub.title,
        body=sub.body,
        status=sub.status,
        reviewer_id=sub.reviewer_id,
        created_at=sub.created_at,
        updated_at=sub.updated_at,
    )


class CreateRequest(BaseModel):
    content_type: ContentType
    title: str
    body: dict[str, Any] = {}


class UpdateRequest(BaseModel):
    title: Optional[str] = None
    body: Optional[dict[str, Any]] = None


class ReviewRequest(BaseModel):
    verdict: ReviewVerdict
    comment: Optional[str] = None


# ── Creator endpoints ─────────────────────────────────────────────────────────

@router.post("/", response_model=SubmissionOut)
def create_endpoint(
    body: CreateRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_platform_role("creator")),
) -> SubmissionOut:
    return _sub_out(create_submission(db, current_user.id, body.content_type, body.title, body.body))


@router.get("/mine", response_model=list[SubmissionOut])
def list_mine(
    status: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[SubmissionOut]:
    return [_sub_out(s) for s in list_submissions(db, author_id=current_user.id, status=status)]


@router.get("/{submission_id}", response_model=SubmissionOut)
def get_endpoint(
    submission_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> SubmissionOut:
    sub = get_submission_or_404(db, submission_id)
    if sub.author_id != current_user.id and not current_user.is_admin:
        roles: list[str] = current_user.platform_roles or []
        if "reviewer" not in roles and "publisher" not in roles:
            from app.core.exceptions import forbidden_exception
            raise forbidden_exception
    return _sub_out(sub)


@router.patch("/{submission_id}", response_model=SubmissionOut)
def update_endpoint(
    submission_id: int,
    body: UpdateRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> SubmissionOut:
    sub = get_submission_or_404(db, submission_id)
    if sub.author_id != current_user.id:
        from app.core.exceptions import forbidden_exception
        raise forbidden_exception
    return _sub_out(update_submission(db, sub, body.title, body.body))


@router.post("/{submission_id}/submit", response_model=SubmissionOut)
def submit_endpoint(
    submission_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> SubmissionOut:
    sub = get_submission_or_404(db, submission_id)
    if sub.author_id != current_user.id:
        from app.core.exceptions import forbidden_exception
        raise forbidden_exception
    return _sub_out(submit_for_review(db, sub))


# ── Reviewer endpoints ────────────────────────────────────────────────────────

@router.get("/queue/pending", response_model=list[SubmissionOut])
def pending_queue(
    db: Session = Depends(get_db),
    _: User = Depends(require_platform_role("reviewer")),
) -> list[SubmissionOut]:
    return [_sub_out(s) for s in list_submissions(db, statuses=["pending_review", "in_review"])]


@router.post("/{submission_id}/claim", response_model=SubmissionOut)
def claim_endpoint(
    submission_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_platform_role("reviewer")),
) -> SubmissionOut:
    sub = get_submission_or_404(db, submission_id)
    return _sub_out(claim_review(db, sub, current_user.id))


@router.post("/{submission_id}/review", response_model=dict[str, Any])
def review_endpoint(
    submission_id: int,
    body: ReviewRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_platform_role("reviewer")),
) -> dict[str, Any]:
    sub = get_submission_or_404(db, submission_id)
    review = record_review(db, sub, current_user.id, body.verdict, body.comment)
    return {
        "review_id": review.id,
        "verdict": review.verdict.value,
        "submission_status": sub.status.value,
    }


# ── Publisher endpoints ───────────────────────────────────────────────────────

@router.get("/queue/approved", response_model=list[SubmissionOut])
def approved_queue(
    db: Session = Depends(get_db),
    _: User = Depends(require_platform_role("publisher")),
) -> list[SubmissionOut]:
    return [_sub_out(s) for s in list_submissions(db, status="approved")]


@router.post("/{submission_id}/publish", response_model=SubmissionOut)
def publish_endpoint(
    submission_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_platform_role("publisher")),
) -> SubmissionOut:
    sub = get_submission_or_404(db, submission_id)
    return _sub_out(publish_submission(db, sub, current_user.id))

File: platform/frontend/lib/api/content.ts
import { apiClient } from "./client";

export type ContentType = "challenge" | "lab";
export type ContentStatus =
  | "draft" | "pending_review" | "in_review" | "approved" | "published" | "rejected";
export type ReviewVerdict = "approved" | "rejected" | "needs_changes";

export interface FlagInput {
  value: string;
  flag_type: "static" | "dynamic" | "per_user" | "per_team";
  case_sensitive: boolean;
}

export interface HintInput {
  order_num: number;
  content: string;
  cost: number;
}

export interface ContentSubmission {
  id: number;
  author_id: number;
  author_username: string | null;
  content_type: ContentType;
  title: string;
  body: Record<string, unknown>;
  status: ContentStatus;
  reviewer_id: number | null;
  created_at: string;
  updated_at: string;
}

export async function getMySubmissions(status?: ContentStatus): Promise<ContentSubmission[]> {
  const { data } = await apiClient.get<ContentSubmission[]>("/content/mine", {
    params: status ? { status } : undefined,
  });
  return data;
}

export async function createSubmission(payload: {
  content_type: ContentType;
  title: string;
  body: Record<string, unknown>;
}): Promise<ContentSubmission> {
  const { data } = await apiClient.post<ContentSubmission>("/content/", payload);
  return data;
}

export async function updateSubmission(
  id: number,
  payload: { title?: string; body?: Record<string, unknown> }
): Promise<ContentSubmission> {
  const { data } = await apiClient.patch<ContentSubmission>(`/content/${id}`, payload);
  return data;
}

export async function submitForReview(id: number): Promise<ContentSubmission> {
  const { data } = await apiClient.post<ContentSubmission>(`/content/${id}/submit`);
  return data;
}

export async function getPendingQueue(): Promise<ContentSubmission[]> {
  const { data } = await apiClient.get<ContentSubmission[]>("/content/queue/pending");
  return data;
}

export async function getApprovedQueue(): Promise<ContentSubmission[]> {
  const { data } = await apiClient.get<ContentSubmission[]>("/content/queue/approved");
  return data;
}

export async function claimSubmission(id: number): Promise<ContentSubmission> {
  const { data } = await apiClient.post<ContentSubmission>(`/content/${id}/claim`);
  return data;
}

export async function reviewSubmission(
  id: number,
  verdict: ReviewVerdict,
  comment?: string
): Promise<{ review_id: number; verdict: ReviewVerdict; submission_status: ContentStatus }> {
  const { data } = await apiClient.post(`/content/${id}/review`, { verdict, comment });
  return data;
}

export async function publishSubmission(id: number): Promise<ContentSubmission> {
  const { data } = await apiClient.post<ContentSubmission>(`/content/${id}/publish`);
  return data;
}
